### DISCUSSION TITLE
Find tests freezing test runner

### DISCUSSION THREAD

**Discussion by dilandau2001** [xunit/xunit#2445] [Question] [answered]
hello.

I have seen similar questions either here and StackOverflow but I would like to condense and explain myself.

### **Context**
I am working with a Visual Studio 2019 solution, running on windows, that already have around 6 thousand tests. And we were having issues with some of them. It is difficult to find culprit as some of them are totally hanging randomly the test host.
We are already using:
    "diagnosticMessages": true,
    "longRunningTestSeconds" :  30
But that is not enough.
Not all test has already injected ITestOutputHelper, I would say less than 10% have it. But traces in a pararell test is not helping either.

Xunit version is: xunit 2.4.1
Visual Studio runner version is: 2.4.3
We pass test using: vstest.console.exe with /Pararell /Platform:x64 and /Enablecoverage

So our current approach right now is:

1. Find a way to make traces aware of the test they are running on
2. Make all testes print Start and Stop traces.
3. Make it "as much transparent as possible". I mean not to modify all test clases and all WriteLines individually.

### **Approach**
**1- Find a way to make traces aware of the test they are running on** 

For this what I have done so far is to write an extension method that uses reflection to get the test name and format the result. I leave it here for reference. 

```
public static class Extension
    {
        /// <summary>Adds a line of text to the output.</summary>
        /// <param name="logger"></param>
        /// <param name="message">The message</param>
        public static void WriteLineExtension(this ITestOutputHelper logger, string message)
        {
            // Extract test name
            var helper = (TestOutputHelper)logger;

            ITest test = (ITest)helper.GetType().GetField("test", BindingFlags.NonPublic | BindingFlags.Instance)
                ?.GetValue(helper);

            var testName = test?.DisplayName;

            var trace = $"[{testName}] - [N] - {message}]";

            logger.WriteLine(message);
        }

        /// <summary>
        /// Writes the line.
        /// </summary>
        /// <param name="logger">The logger.</param>
        /// <param name="format">The format.</param>
        /// <param name="args">The arguments.</param>
        public static void WriteLineExtension(this ITestOutputHelper logger, string format, params object[] args)
        {
            // Extract test name
            var helper = (TestOutputHelper)logger;

            ITest test = (ITest)helper.GetType().GetField("test", BindingFlags.NonPublic | BindingFlags.Instance)
                ?.GetValue(helper);

            var testName = test?.DisplayName;

            var trace = $"[{testName}] - [N] - {format}]";

            logger.WriteLine(trace, args);
        }
    }
```

This way I am able to accomplish what I need but it requires to modify (just a text replace in the solution) all traces.

_**Is there any another way to really override default ITestOutputHelper behaviour so I don't really need to modify all traces in the code? Can I decorate TestOutputHelper and modify which instance of the helper is injected in the test?**_ 

**2- Make all testes print Start and Stop traces.** 

For this I have thought 3 possible ways. Although one is not working and all of them implies too many code changes I would like to avoid.

Firts approach and working one is to do a base class.

```
    /// <summary>
    /// Abstrac class to print when every test is started and ended.
    /// </summary>
    public abstract class TestsBase : IDisposable
    {
        private ITestOutputHelper _helper;

        protected TestsBase(ITestOutputHelper helper)
        {
            _helper = helper;
            _helper.WriteLineExtension("TEST START");
        }

        public void Dispose()
        {
            _helper.WriteLineExtension("TEST END");
        }
    }
```

this works but again requires me to modify all test classes to inherit from it and to pass a ITestOutputHelper that might be not already injected in the code. This is also a problem when the existing test class already inherits from a base class.
I am trying to avoid making too many modifications if there is a clever way to make it generally all around with the less changes possible.

Another approach I tried but it did not work, was to make a classFixture like this:

```
    public class TestsBaseFixture : IAsyncLifetime
    {
        public ITestOutputHelper Helper { get; set; }

        public TestsBaseFixture(ITestOutputHelper helper)
        {
            Helper = helper;
        }

        public Task DisposeAsync()
        {
            Helper.WriteLineExtension("TEST END");
            return Task.CompletedTask;
        }

        public Task InitializeAsync()
        {
            Helper.WriteLineExtension("TEST START");
            return Task.CompletedTask;
        }
    }
```

This did not work because class fixtures don't support ITestOutputHelper. **_Can I "get" a valid ITestOutputHelper anywhere else in the code?_** I mean, is magically being injected by the constructor the only possible way to have ITestOutputHelper or is there a factory or similar that I could call to retrieve the current valid ITestOutputHelper (so traces are attached to the test in the report)?. If there is a way to get the ITestOutputHelper then this approach could work just by creating it in the constructor instead of getting the injected one.
This approach would have require also to modify all classes but with less changes as I dont have to modify existing test class constructors. Another caveat is that not sure if this works when test already in a CollectionFixture.

Another approach I have think of, I guess could be done but I don't know how to do it is to somehow override [Fact] and [Theory] and somehow make them aware of the ITestOutputHelper. Don't know if this is possible.

Also I read in some old post for Xunit Test runner, a -showProgress configuration that was already implemented doing what I wanted for test progress. I don't know if there is something similar in xunit for visual studio. 

**_Is there any other way to do it?_**

### Summary of questions
If you reach this, congratulations. I know it is a big post but I wanted to make it as complete as possible.
I have already send many questions but all them are related to the main issue to identify test that are hunging the test runner. 

My current only valid functional approach is the TestBase class and the WriteLine extension. This would work but require many code modifications.

Do you think that is the correct approach? Can you come up with any other approach I haven't try yet?
Thanks

**Comment by bradwilson** [comment, marked as answer]
>Can I decorate TestOutputHelper and modify which instance of the helper is injected in the test?

Doing this would involved a very large amount of extensibility work on your part, at which point you could inject anything you wanted to (not just a custom `ITestOutputHelper` instance).

>Can I "get" a valid ITestOutputHelper anywhere else in the code?

No. As its name implies, it's only valid for a single test.

Fixtures live a life outside of a single test. They can take [`IMessageSink`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Messages/BaseInterfaces/IMessageSink.cs) as a constructor argument, and send instances of [`IDiagnosticMessage`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Messages/IDiagnosticMessage.cs) to it, which will be shown when diagnostic messages are enabled. So a fixture that was passed to a test could provide a trace method that emitted diagnostic messages, which should be shown immediately (no need to wait for the output helper, which collects text internally while waiting for the test to complete).

Additionally, you can use the [Caller Information attributes](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/caller-information) on this method to automatically obtain the calling method name, filename, and/or line number, which presumably would be enough information to help you determine where things are going sideways.

**Reply by dilandau2001** [reply]
First of all. Thanks for the answer. 

> Doing this would involved a very large amount of extensibility work on your part, at which point you could inject anything you wanted to (not just a custom `ITestOutputHelper` instance).

Sad. I was hoping there would be some easy way to access something like a container (unity or autofac or similar) in order to modify the default instance of TestOutputHelper that is latetely injected on each test.

Or a way to modify "writeLine" behaviour of current TestOutPutHelper so it always prints test name before the message. (or whatever extra information and format we would be needing)


> They can take [`IMessageSink`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Messages/BaseInterfaces/IMessageSink.cs) as a constructor argument, and send instances of [`IDiagnosticMessage`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Messages/IDiagnosticMessage.cs) to it, which will be shown when diagnostic messages are enabled. So a fixture that was passed to a test could provide a trace method that emitted diagnostic messages, which should be shown immediately (no need to wait for the output helper, which collects text internally while waiting for the test to complete).

So. If I understand it correctly. I can inject IMessageSink to the Fixture contructor and I can implement a Trace function in the fixture that can be called from each test. Would be these traces related to the running test?. This is not very far from just using current instance of ITraceOutputHelper.


What I want to achieve is not difficult If I was willing to modify all 6 thousand tests, and all testoutput Writelines in the solution. But this is something I was investigating how to avoid.

Ideally I would like to modify [Fact] and [Theory] default behaviours so they magically print Test Start/Test End for each test and modify Writeline format of all traces within each test so it is possible to identify test parent when console flat text is analyzed (I mean the text from a running test console or Tfs log). If this is posible I would be achieving the result without the need to modify any of the 6000 tests. Somehow this is known by the tfs when the test results are summarized but I don't see this relation when raw log is opened.

**Reply by bradwilson** [reply]
>So. If I understand it correctly. I can inject IMessageSink to the Fixture contructor and I can implement a Trace function in the fixture that can be called from each test. Would be these traces related to the running test?

No, these would be sent as diagnostic messages, without any relation to any specific test. That's why I suggested you use the extra parameters with the attributes that would allow you to prepend the diagnostic message with information about where it came from (method name, filename, line number).

>What I want to achieve is not difficult If I was willing to modify all 6 thousand tests, and all testoutput Writelines in the solution. But this is something I was investigating how to avoid.

Unfortunately there's no simple way to do this today because `ITestOutputHelper` does not have any extensibility points, without the aforementioned modification of the test execution pipeline. The instance that's chosen to fill the injection of `ITestOutputHelper` is [done in XunitTestClassRunner](https://github.com/xunit/xunit/blob/59dea5bcc7048e666a6b0e317173dd80a8f72b76/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestClassRunner.cs#L185-L196), which means you'd need to replace at least that much of the execution pipeline to override the `TryGetConstructorArgument` function. There is an example of doing this in the [Observation sample project](https://github.com/xunit/samples.xunit/tree/main/ObservationExample), though it's more than you need, since it's entirely replacing the execution pipeline. You'd want to just make a few surgical changes to the pipeline, inheriting from the `Xunit*` runners (assembly, test collection, and test class) in order to get your one override in place, and then you could continue to use the existing pipeline (letting your custom class runner hand off directly to `XunitTestMethodRunner`).

This would definitely be the least intrusive way to change the output helper, though it doesn't necessarily help the problem that output is saved until you finish executing. You could overcome this by having an output helper that actually just turned the messages into diagnostic messages (the class runner has access to the diagnostic message sink, so it could give that instance to your custom output helper).
